lib/lambda_ethereum_consensus/validator/validator_set.ex

```diff
@@ -0,0 +1,284 @@
+  def notify_tick(%{head_root: head_root} = set, {slot, third} = slot_data) do
+    Logger.debug("[ValidatorSet] Tick #{inspect(third)}", root: head_root, slot: slot)
+    epoch = Misc.compute_epoch_at_slot(slot)
+
+    set
+    |> update_state(epoch, slot, head_root)
+    |> process_tick(epoch, slot_data)
+  end
+
+  defp process_tick(%{head_root: head_root} = set, epoch, {slot, :first_third}) do
+    propose(set, epoch, slot, head_root)
+  end
+
+  defp process_tick(%{head_root: head_root} = set, epoch, {slot, :second_third}) do
+    set
+    |> attests(epoch, slot, head_root)
+    |> build_payload(slot + 1, head_root)
+  end
+
+  defp process_tick(set, epoch, {slot, :last_third}) do
+    publish_aggregates(set, epoch, slot)
+  end
+
+  ##############################
+  # State update
+
+  defp update_state(set, epoch, slot, head_root) do
+    set
+    |> update_head(head_root)
+    |> compute_duties(epoch, slot, head_root)
+  end
+
+  defp update_head(%{head_root: head_root} = set, head_root), do: set
+  defp update_head(set, head_root), do: %{set | head_root: head_root}
+
+  defp compute_duties(set, epoch, _slot, _head_root)
+       when is_duties_computed(set, epoch) and is_duties_computed(set, epoch + 1),
+       do: set
+
+  defp compute_duties(set, epoch, slot, head_root) do
+    epochs_to_calculate =
+      [{epoch, slot}, {epoch + 1, Misc.compute_start_slot_at_epoch(epoch + 1)}]
+      |> Enum.reject(&Map.has_key?(set.duties, elem(&1, 0)))
+
+    epochs_to_calculate
+    |> Map.new(&compute_duties_for_epoch!(set, &1, head_root))
+    |> merge_duties_and_prune(epoch, set)
+  end
+
+  defp compute_duties_for_epoch!(set, {epoch, slot}, head_root) do
+    beacon = Validator.fetch_target_state_and_go_to_slot(epoch, slot, head_root)
+
+    duties = %{
+      proposers: Duties.compute_proposers_for_epoch(beacon, epoch, set.validators),
+      attesters: Duties.compute_attesters_for_epoch(beacon, epoch, set.validators)
+    }
+
+    Duties.log_duties_for_epoch(duties, epoch)
+
+    {epoch, duties}
+  end
+
+  defp merge_duties_and_prune(new_duties, current_epoch, set) do
+    set.duties
+    # Remove duties from epoch - 2 or older
+    |> Map.reject(fn {old_epoch, _} -> old_epoch < current_epoch - 1 end)
+    |> Map.merge(new_duties)
+    |> then(fn current_duties -> %{set | duties: current_duties} end)
+  end
+
+  ##############################
+  # Block proposal
+
+  defp build_payload(%{validators: validators} = set, slot, head_root) do
+    # We calculate payloads from a previous slot, we need to recompute the epoch
+    epoch = Misc.compute_epoch_at_slot(slot)
+
+    case Duties.current_proposer(set.duties, epoch, slot) do
+      nil ->
+        set
+
+      validator_index ->
+        validators
+        |> Map.update!(validator_index, &Validator.start_payload_builder(&1, slot, head_root))
+        |> update_validators(set)
+    end
+  end
+
+  defp propose(%{validators: validators} = set, epoch, slot, head_root) do
```

Comment:
nit `maybe_propose`?

Reply:
Sure, should do the same with attest and build payload

Reply:
Done